### ISSUE TITLE
Regression false-positive FinalLocalVariable

### ISSUE THREAD

**Issue by archmageirvine** [checkstyle/checkstyle#3172]
The variable `s` in the following code cannot be made final, yet recent versions of checkstyle claims that it can.

$ java -jar ~/java/cvscheck/lib/checkstyle-6.18-all.jar -c myconfig.xml Bug.java
Starting audit...
[WARN] /home/sean/Bug.java:13:15: Variable 's' should be declared final. [FinalLocalVariable]
Audit done.

``` java
import java.util.Arrays;
import java.util.ArrayList;

public final class Bug {

  Bug() { }

  private static boolean f() {
    return Math.random() > 0;
  }

  public static void main(final String[] args) {
    Integer[] s;
    if (f()) {
      s = new Integer[0];
    } else {
      s = new Integer[5];
      if (f()) {
        final ArrayList<Integer> retain = new ArrayList<>();
        for (final Integer c : s) {
          if (f()) {
            retain.add(c);
          }
        }
        s = retain.toArray(new Integer[retain.size()]);
      }
      if (f() && s.length > 0) {
        s = new Integer[] {s[0]};
      }
    }
    System.out.println(Arrays.toString(s));
  }
}
```

Content of config:

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE module PUBLIC "-//Puppy Crawl//DTD Check Configuration 1.2//EN" "http://www.puppycrawl.com/dtds/configuration_1_2.dtd">
<module name="Checker">
    <property name="severity" value="warning"/>
    <module name="TreeWalker">
        <module name="FinalLocalVariable">
            <property name="tokens" value="VARIABLE_DEF,PARAMETER_DEF"/>
        </module>
    </module>
</module>
```
##

**Comment by romani** [comment]
Do you know in what checktyle version, there were no violation?

**Comment by archmageirvine** [comment]
It was fine in 6.1 (and possibly later).

**Comment by rnveach** [comment]
6.14 was the first CS version that showed no violations.

**Comment by romani** [comment]
@Luolc , can you help us with this issue ?

**Comment by Luolc** [comment]
@romani I will look over the related codes and get familiar with this issue. Currently I am focusing on  the indentation issue. So I might not be able to solve this immediately.

**Comment by romani** [comment]
this is not easy issue , you will not do it immediately for sure.

**Comment by romani** [comment]
fix is merged
